### ISSUE TITLE
On Linux: respect XDG_CONFIG_HOME and don't store config in user's home directory

### ISSUE THREAD

**Issue by sethwoodworth** [sqlfluff/sqlfluff#405]
The XDG Base Directory standard states that user config files should respect the `XDG_CONFIG_HOME` environment variable, which defaults `$HOME/.config`.

Currently, sqlfluff looks for a user's global config file in `~/.sqlfluff`.  I, like many heavy terminal users, have a *lot* of files in our home directories, and prefer to use the `~/.config` directory whenever possible.

The `.config` dir is only relevant to Linux users.  On OSX, the standard says config files should live in ` ~/Library/Application Support/<AppName>`.  On Windows the standard is `C:\Documents and Settings\<User>\Application Data\Local Settings\<AppAuthor>\<AppName>`. The [`appdirs` pypi package](https://pypi.org/project/appdirs/) makes all three of these standards simple to implement. And I suggest providing a fallback to `~/.sqlfluff` should  there not be an OS specific sqlfluff config file present.

If you agree with this issue, I'd be happy to implement it!

**Comment by alanmcruickshank** [comment]
Hi @sethwoodworth - I wasn't aware of that package - looks great!

Yes if you're happy to implement, I'd be very happy to receive a PR on this.

We probably have some existing users who want to keep their config in `~/.sqlfluff`, so I'd love to keep that as an option. You'll see in that the logic for the config files means that they layer on top of each other, so the ordering matters, and I think for config files in the locations that you mention that they should come *before* any files found at `~/.sqlfluff`.

In other words, if you look at the docs here: https://docs.sqlfluff.com/en/latest/configuration.html#nesting, I think that the config files in the locations you mention should come between step 0 and step 1.

Finally if you could update the docs page that I've linked above in the same PR describing what you've done, that would be excellent. I'm looking forward to this one!

**Comment by sethwoodworth** [comment]
@alanmcruickshank Wonderful!
I started implementing the feature yesterday, and implemented it as replacing the result of `load_user_config` if the directory exists on the user's system.  My thought was that a user would have one user config, but I remember now looking at the nesting docs that you can have multiple files in the same directory that are apply configs, so I'll be sure to implement the appdirs version that way as well.

How would you like me to test this?  My go-to solution would be to bring in pytest and to mock the `appdirs` library, but you're not doing any mocking in `config_test.py` and `load_user_config` isn't currently tested.

**Comment by alanmcruickshank** [comment]
Excellent question on testing this.

You're correct that it's not properly tested at the moment. If you feel happy to put together something which mocks `appdirs` then that would be great to get some better test coverage in that part of the codebase. If you feel like covering all of the config loading routines that would be great - otherwise it would still be useful if you put some testing in for yours and pointed toward how someone else could cover the others 👍 .

thanks again for putting together the PR. Looks like great work.

**Comment by sethwoodworth** [comment]
The matching PR was merged, do you want to close this issue, or keep it open until tests are provided?

**Comment by NiallRees** [comment]
Let's close for now as the initial issue is resolved. If there's a need for a new issue let's do that though.
